gslib/commands/rpo.py

```diff
@@ -0,0 +1,213 @@
+
+_SET_DESCRIPTION = """
+<B>SET</B>
+  The ``rpo set`` command configures turbo replication
+  for dual-region Google Cloud Storage buckets.
+
+<B>SET EXAMPLES</B>
+  Configure your buckets to use turbo replication:
+
+    gsutil rpo set ASYNC_TURBO gs://redbucket gs://bluebucket
+
+  Configure your buckets to NOT use turbo replication:
+
+    gsutil rpo set DEFAULT gs://redbucket gs://bluebucket
+"""
+
+_GET_DESCRIPTION = """
+<B>GET</B>
+  The ``rpo get`` command returns the replication setting
+  for the specified Cloud Storage buckets.
+
+<B>GET EXAMPLES</B>
+  Check if your buckets are using turbo replication:
+
+    gsutil rpo get gs://redbucket gs://bluebucket
+"""
+
+_DESCRIPTION = """
+  The ``rpo`` command is used to retrieve or configure the
+  `replication setting
+  <https://cloud.google.com/storage/docs/turbo-replication>`_ of
+  dual-region Cloud Storage buckets.
+  This command has two sub-commands: ``get`` and ``set``.
+""" + _GET_DESCRIPTION + _SET_DESCRIPTION
+
+_DETAILED_HELP_TEXT = CreateHelpText(_SYNOPSIS, _DESCRIPTION)
+_set_help_text = CreateHelpText(_SET_SYNOPSIS, _SET_DESCRIPTION)
+_get_help_text = CreateHelpText(_GET_SYNOPSIS, _GET_DESCRIPTION)
+
+
+class RpoCommand(Command):
+  """Implements the gsutil rpo command."""
+
+  command_spec = Command.CreateCommandSpec(
+      'rpo',
+      command_name_aliases=[],
+      usage_synopsis=_SYNOPSIS,
+      min_args=2,
+      max_args=NO_MAX,
+      supported_sub_args='',
+      file_url_ok=False,
+      provider_url_ok=False,
+      urls_start_arg=2,
+      gs_api_support=[ApiSelector.JSON],
+      gs_default_api=ApiSelector.JSON,
+      argparse_arguments={
+          'get': [CommandArgument.MakeNCloudURLsArgument(1),],
+          'set': [
+              CommandArgument('mode', choices=list(VALID_RPO_VALUES)),
+              CommandArgument.MakeZeroOrMoreCloudBucketURLsArgument()
+          ],
+      })
+  # Help specification. See help_provider.py for documentation.
+  help_spec = Command.HelpSpec(
+      help_name='rpo',
+      help_name_aliases=[],
+      help_type='command_help',
+      help_one_line_summary='Configure replication',
+      help_text=_DETAILED_HELP_TEXT,
+      subcommand_help_text={
+          'get': _get_help_text,
+          'set': _set_help_text,
+      },
+  )
+
+  def _ValidateBucketListingRefAndReturnBucketName(self, blr):
+    if blr.storage_url.scheme != 'gs':
+      raise CommandException(
+          'The %s command can only be used with gs:// bucket URLs.' %
+          self.command_name)
+
+  def _GetRpo(self, blr):
+    """Gets the rpo setting for a bucket."""
+    bucket_url = blr.storage_url
+
+    bucket_metadata = self.gsutil_api.GetBucket(bucket_url.bucket_name,
+                                                fields=['rpo'],
+                                                provider=bucket_url.scheme)
+    rpo = bucket_metadata.rpo
+    bucket = str(bucket_url).rstrip('/')
+    print('%s: %s' % (bucket, rpo))
```

Comment:
If `rpo` is a raw Apitools values, it might look bad. For Autoclass, it looked like:
```
<AutoclassValue enabled=True,
    toggleTime=<DateTimeValue ... >
```

Reply:
That is how the json api was generated by the backend. We can't do much here.

Reply:
What do you mean? Can't you print the relevant properties of the object as strings instead of directly printing the object?

Reply:
See gslib/third_party/storage_apitools/storage_v1_messages.py. rpo is a StringField and not a Class like autoclassValue. Hence we can print it directly.